### PR TITLE
[operator-trivy] add insecure db registry and non-ssl registries

### LINKED ISSUE #6608: [operator-trivy] add http schema option to trivy db downloader
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Use case. Why is this important?

Some clients may run a deckhouse-operated cluster in an air-gapped environment providing access to trivy db only over http.

### Proposed Solution

Figure out if OCI package (used by trivy to download DB) supports downloading artifacts via http and implement necessary changes. 

### Additional Information

```
[root@box3-k8s-m1 api_gateway]# kubectl -n d8-operator-trivy logs trivy-server-0
2023-11-20T07:39:50.650Z        INFO    Need to update DB
2023-11-20T07:39:50.650Z        INFO    DB Repository: [box3-pgmain-s1.imp.XXX.io:8001/deckhouse/security/trivy-db](http://box3-pgmain-s1.imp.XXX.io:8001/deckhouse/security/trivy-db)
2023-11-20T07:39:50.650Z        INFO    Downloading DB...
2023-11-20T07:39:50.657Z        FATAL   failed to download vulnerability DB: database download error: OCI repository error: 1 error occurred:
        * Get "https://box3-pgmain-s1.imp.XXX.io:8001/v2/": http: server gave HTTP response to HTTPS client
```

### LINKED ISSUE #10471: [operator-trivy] download trivy db from insecure registry
### Preflight Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/deckhouse/deckhouse/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/deckhouse/deckhouse/issues) for an issue that matches the one I want to file, without success.

### Version

v1.65.0

### Expected Behavior

- It should be possible to download Trivy DB from an insecure registry (using http instead of https, look the description of actual behavior below);
- It makes sense to make `trivy.dbRepositoryInsecure` setting customizable.

### Actual Behavior

It seems, despite the fact that `trivy.dbRepositoryInsecure` is set to "true", Trivy tries to download its database from an insecure registry using HTTPS  in case the registry host is specified with a port number:
```
2024-11-05T11:08:24Z    FATAL    Fatal error    failed to download vulnerability DB: OCI artifact error: failed to download vulnerability DB: failed to download artifact from nginx.default.svc.cluster.local:80/trivy-db:2: OCI repository error: 1 error occurred:
    * Get "https://nginx.default.svc.cluster.local:80/v2/": http: server gave HTTP response to HTTPS client; GET http://nginx.default.svc.cluster.local:80/v2/: unexpected status code 404 Not Found: <html>
<head><title>404 Not Found</title></head>
```

### Steps To Reproduce

1. set trivy.dbRepositoryInsecure to true
2. set dbRepository settings in the configmap to some address backed by an http web-server
3. make trivy-server download trivy-db

### Additional Information

_No response_

### Logs

_No response_

### PR DESCRIPTION
**Pull request by miklezzzz**
## Description
Provides a new setting that allows downloading databases from insecure registries.

## Why do we need it, and what problem does it solve?
Some users may want to use the operator in an air-gapped environment, hosting images and databases in insecure registries.
Closes #10471 
Closes #6608 

## What is the expected result?
Trivy-server is able to download databases/images from insecure registries.

## Checklist
- [x] The code is covered by unit tests.
- [ ] e2e tests passed.
- [x] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: operator-trivy
type: chore
summary: Add a setting for downloading images from insecure registries.
impact_level: default
```

### CODE REVIEW

**Comment by miklezzzz** [comment]
```
spec:
  enabled: true
  settings:
    nonSslRegistries:
    - my.example.com:8888
    - test.com
```

```
  TRIVY_INSECURE: "false"
  trivy.dbRepositoryInsecure: "false"
  trivy.nonSslRegistry.0: my.example.com:8888
  trivy.nonSslRegistry.1: test.com
```

**Comment by miklezzzz** [comment]
```
spec:
  enabled: true
  settings:
    insecureDbRegistry: true
```
```
  TRIVY_INSECURE: "true"
  trivy.dbRepositoryInsecure: "true"
```

**Comment by miklezzzz** [comment]
insecureDbRegistry: false
```
2024-11-10T09:19:46Z	INFO	[vulndb] Downloading vulnerability DB...
2024-11-10T09:19:46Z	INFO	[vulndb] Downloading artifact...	repo="nexus.dev.m-scherba.hf.flant.com:8088/repository/security/trivy-db:2"
2024-11-10T09:19:46Z	FATAL	Fatal error	failed to download vulnerability DB: OCI artifact error: failed to download vulnerability DB: failed to download artifact from nexus.dev.m-scherba.hf.flant.com:8088/repository/security/trivy-db:2: OCI repository error: 1 error occurred:
	* Get "https://nexus.dev.m-scherba.hf.flant.com:8088/v2/": http: server gave HTTP response to HTTPS client
```

**Comment by miklezzzz** [comment]
insecureDbRegistry: true
```
2024-11-10T09:20:26Z	INFO	[vulndb] Downloading vulnerability DB...
2024-11-10T09:20:26Z	INFO	[vulndb] Downloading artifact...	repo="nexus.dev.m-scherba.hf.flant.com:8088/repository/security/trivy-db:2"
2024-11-10T09:20:35Z	INFO	[vulndb] Artifact successfully downloaded	repo="nexus.dev.m-scherba.hf.flant.com:8088/repository/security/trivy-db:2"
2024-11-10T09:20:35Z	INFO	Listening 0.0.0.0:4954...
```

File: ee/modules/500-operator-trivy/openapi/doc-ru-config-values.yaml [hunk 1]
```diff
@@ -48,3 +52,7 @@ properties:
   insecureRegistries:
     description: |
       Хранилище образов контейнеров (container registry), к которым разрешены незащищённые соединения.
```
Comment (z9r5) [inline ee/modules/500-operator-trivy/openapi/doc-ru-config-values.yaml:54]:
```suggestion
      Список адресов хранилищ образов контейнеров (container registry), к которым разрешены недоверенные HTTPS-подключения (не прошедшие проверку TLS-сертификата).
      
      > Для работы необходимо включение параметра [insecureDbRegistry](#parameters-insecuredbregistry).
```

File: ee/modules/500-operator-trivy/openapi/doc-ru-config-values.yaml [hunk 2]
```diff
@@ -48,3 +52,7 @@ properties:
...
+
+  nonSslRegistries:
+    description: |
+      Хранилище образов контейнеров (container registry), к которым разрешены незашифрованные (non-ssl) подключения.
```
Comment (z9r5) [inline ee/modules/500-operator-trivy/openapi/doc-ru-config-values.yaml:58]:
```suggestion
      Список адресов хранилищ образов контейнеров (container registry), к которым разрешены подключения по HTTP.
      
      > Для работы необходимо включение параметра [insecureDbRegistry](#parameters-insecuredbregistry).
```

File: ee/modules/500-operator-trivy/openapi/config-values.yaml [hunk 3]
```diff
@@ -101,10 +109,25 @@ properties:
...
-    x-doc-example: [["my.registry.com", "http-only.registry.io"]]
+    x-examples:
+    -
+      - my.registry.com
+      - http-only.registry.io
     items:
       type: string
     description: |
       Container registries to which insecure connections are allowed.
+
```
Comment (z9r5) [inline ee/modules/500-operator-trivy/openapi/config-values.yaml:123]:
```suggestion
      List of container registry addresses to which insecure HTTPS connections are allowed (not passed TLS certificate verification).

      > The parameter [insecureDbRegistry](#parameters-insecuredbregistry) must be enabled.
```

**Review by z9r5 — COMMENTED** [review]
Some suggestions

File: ee/modules/500-operator-trivy/templates/configmap.yaml [hunk 4]
```diff
@@ -62,11 +62,20 @@ data:
   {{- range $idx, $registry := .Values.operatorTrivy.insecureRegistries }}
   trivy.insecureRegistry.{{ $idx }}: {{ $registry | quote }}
```
Comment (yalosev) [inline ee/modules/500-operator-trivy/templates/configmap.yaml:63]:
```suggestion
  trivy.insecureRegistry.{{ $idx }}: {{ $registry | quote }}
  trivy.nonSslRegistry.{{ $idx }}: {{ $registry | quote }}
```

File: ee/modules/500-operator-trivy/openapi/config-values.yaml [hunk 5]
```diff
@@ -101,10 +109,25 @@ properties:
...
+  nonSslRegistries:
```
Comment (yalosev) [inline ee/modules/500-operator-trivy/openapi/config-values.yaml:124]: Include this logic into the`insecureRegistries` option

Comment (miklezzzz) [reply ee/modules/500-operator-trivy/openapi/config-values.yaml]: done

File: ee/modules/500-operator-trivy/templates/configmap.yaml (hunk 4 above)
Comment (miklezzzz) [reply ee/modules/500-operator-trivy/templates/configmap.yaml:63]: done

File: ee/modules/500-operator-trivy/openapi/config-values.yaml [hunk 6]
```diff
@@ -101,9 +109,13 @@ properties:
     -
       - app
       - env
+
   insecureRegistries:
     type: array
```
Comment (z9r5) [inline ee/modules/500-operator-trivy/openapi/config-values.yaml:114]:
```suggestion
    type: array
    description: |
      List of container registry addresses to which insecure HTTPS connections (not passed TLS certificate verification) or HTTP connections are allowed.

      > The parameter [insecureDbRegistry](#parameters-insecuredbregistry) must be enabled.
```

**Review by z9r5 — APPROVED** [review]

**Review by yalosev — APPROVED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Failure: cherry-pick commit 58c125123f73f3afa0464b3d4b5ff4ec2f5b5ac8 to the branch [release-1.65](https://github.com/deckhouse/deckhouse/tree/release-1.65) failed. See [Job](https://github.com/deckhouse/deckhouse/actions/runs/11795656242) for details.

**Comment by Taior** [comment]
Backported to 1.65 in #10607

**Comment by deckhouse-BOaTswain** [comment]
Cherry pick PR [10666](https://github.com/deckhouse/deckhouse/pull/10666) to the branch [release-1.66](https://github.com/deckhouse/deckhouse/tree/release-1.66) successful!

**Pull request merged**